### DISCUSSION TITLE
Arabic language support

### DISCUSSION THREAD

**Discussion by Hadryan** [meilisearch/product#139] [Feedback & Feature Proposal]
> All varieties of Arabic combined are spoken by perhaps as many as 422 million speakers (native and non-native) in the Arab world, making it the fifth most spoken language in the world. (Wikipedia)

Related: https://github.com/meilisearch/MeiliSearch/issues/553

**Comment by Kerollmops** [comment]
Hey,

As I [answered to your previous language support request](https://github.com/meilisearch/MeiliSearch/issues/553#issuecomment-605130705), I would like some feedback on the current state of the engine. We currently split by whitespace and when it comes to CJK character we split them one by one and consider each of those a different word.

Thank you!

**Comment by mzeidhassan** [comment]
Hi @Kerollmops , I think if you use what [Apache Solr](https://lucene.apache.org/solr/guide/6_6/language-analysis.html#LanguageAnalysis-Arabic
) uses, it would be great. 

Character normalization factories are very important.

solr.ArabicNormalizationFilterFactory
solr.ArabicStemFilterFactory
solr.ICUFoldingFilterFactory


So, to have a good Arabic search engine, you need to be able to search by stem. Also, there should be some kind of character normalization, while there is an option for users to get exact searches if they wish,  i.e.: ignoring normalization. Something like quoted search. 

For example, in Arabic, we have alef hamza above, alef hamza below, and alef hamza. Here how hey look like: 

```
أ
إ
ا
```

Character normalization should be able to treat all variations of hamza as the last one. So, hamza above and hamza below can be ignored. Think of it as in German:
Typing "uber", should be able to get 'Über', and 'ueber'. 

Hope this helps.

**Comment by curquiza** [comment]
I add here the comment of @mohamed-foly that could be definitely useful
https://github.com/meilisearch/meilisearch/issues/2248#issuecomment-1081928700

**Reply by curquiza** [reply]
Another interesting related issue: https://github.com/meilisearch/charabia/issues/36

**Comment by curquiza** [comment]
Also, we've just released a new version of our tokenizer, making it easier to contribute. Meaning anyone can add an Arabic normalizer/segmenter and then, make Meilisearch finally compliant with the Arabic language as the Arabic community expects Meilisearch to work 😄 

Feel free to open any PR or even ask any question on the [tokenizer repo](https://github.com/meilisearch/charabia), named charabia 🎉

**Comment by LoaiDev** [comment]
any updates on this?

**Reply by gmourier** [reply]
Hi @LoaiDev, as we don't have any native Arabic-speaking technical person in the core team at the moment, we hope to have an external contribution on the [tokenizer repo](https://github.com/meilisearch/charabia) to move this topic forward. 

We have made contributions easier to provide on the tokenizer lately to favor that language-specific addition. Feel free to contribute; it will be a massive help for Meilisearch. 😊

**Comment by ManyTheFish** [comment]
Hello all!
I come back to you to make some updates.
For [Hacktoberfest](https://hacktoberfest.com/) we've created several issues to enhance the Arabic support of Meilisearch:

- https://github.com/meilisearch/charabia/issues/137:
> A normalization allowing Meilisearch to retrieve words containing nonspacing marks like diacritics or hamza.
- https://github.com/meilisearch/charabia/issues/133: 
> Enhance the Arabic Language support by segmenting agglutinated words like `الشجرة`

All these issues are open to external contributions during the whole month, so don't hesitate to contribute! 🧑‍💻
I'll be glad to help you if you have any questions. 

This is another step in enhancing Arabic Language support, depending on future feedback, we will be able to go further.

Thanks for all your feedback! ✍️

**Comment by ManyTheFish** [comment]
Hello all!

I have a simple question about [Arabic ligatures](https://www.compart.com/en/unicode/search?q=Arabic+Ligature#characters), is it only useful to format texts or does it have a particular sense compared to the decomposed form? For instance `ﰲ` ([U+FC32](https://www.compart.com/en/unicode/U+FC32)) is composed of `ف` ([U+0641](https://www.compart.com/en/unicode/U+0641)) + `ي` ([U+064A](https://www.compart.com/en/unicode/U+064A)), is it sensible to decompose them?

> note that if I write both composing letters together, GitHub seems to automatically add the ligature, making `ﰲ` (U+FC32) and `في` (U+0641) + (U+064A) similar for a non-Arabic writter. 🤔 

Thanks a lot for your help!

**Reply by mohamed-foly** [reply]
@ManyTheFish 
```في```
  a word has a meaning in the sentense , can be translated to ```in``` or ```at``` based on the sentense.
It's the same if you tried to decompose ```in``` to ```i``` and ```n```

**Reply by ManyTheFish** [reply]
Hey @mohamed-foly, thanks for your answer!
I don't suggest splitting these two characters into 2 words, my question is: does ﰲ means the same as في?

To give a bit more context, today, Meilisearch consider them as 2 completely different things.

**Reply by LoaiDev** [reply]
they are the same word. I never knew this even existed

**Reply by LoaiDev** [reply]
if that's the case for all arabic words that means that now meilisearch can only match full words because most letters will change to a different unicode if a letter is written after it

**Reply by mzeidhassan** [reply]
hi @ManyTheFish 
Arabic ligatures are rarely used out of professional liturgical typing. That said, they are indeed 2 different things. In most texts, the composed form is the widely used. As @LoaiDev mentioned, I didn't know these even existed, but apparently there are a [bunch of them](https://www.compart.com/en/unicode/search?q=Arabic+Ligature). 


I haven't seen any Arabic keyboard that supports this form though. 

Maybe, it would be better if you give option to normalize such ligatures, i.e. replacing ﰲ with في

By the way, whenever I copy that ﰲ in Notepad, or any text editor, it shows up as a box (gets corrupted). 

![image](https://user-images.githubusercontent.com/7589948/197558227-dd5235c7-7b24-491b-af2a-b08b4baa3d40.png)

**Reply by ManyTheFish** [reply]
@LoaiDev and @mzeidhassan, thank you for your help.
So the issue about [Compatibility Decomposition (charabia#139)](https://github.com/meilisearch/charabia/issues/139) should enhance Meilisearch Arabic language support by unifying both Unicode versions of Arabic ligatures.

**Comment by DrAliRagab** [comment]
I hope if anyone can help with this: https://github.com/meilisearch/charabia/pull/204

**Comment by hamdallah90** [comment]
any update on this?

**Reply by curquiza** [reply]
Hello @hamdallah90 
The improvement described by @DrAliRagab in [this comment](https://github.com/meilisearch/product/discussions/139#discussioncomment-5538122) is already available in a release candidate ([v1.2.0-rc.2](https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0-rc.2)) and will be release officially in v1.2.0 on 5th of June 😊

**Reply by DrAliRagab** [reply]
Yes, both Arabic normalizer and segmenter are improved.

**Reply by macraig** [reply]
[v1.2.0](https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0) is now officially out and includes normalizer and segmentation improvements.

**Comment by LoaiDev** [comment]
Hello, commenting on this again in case someone can work on it. In arabic there is "ال" which is the same as "the" but it is at the start of the word instead of separate. so "باب" would mean "door" but "الباب" would mean "the door" I am not sure how meilisearch handles "the" in english but I think this also should be handled someway in arabic

**Reply by ManyTheFish** [reply]
Hello @LoaiDev, 
Did you try the last version of Meilisearch? @DrAliRagab worked on this specific case and ensured that "ال" is split from the linked word, if you encounter an issue on this it's probably a bug, do you manage to find a document containing "الباب" by searching "باب"?

Have a good day!

**Reply by lordjoo** [reply]
Actulay we also are facing a similar case where we have "للباب" and try searching with "باب" and no documents found

**Reply by ManyTheFish** [reply]
Interesting, could you send a text that should be found with the query "باب" but isn't retrieved by Meilisearch, please?

**Reply by usamamhmod** [reply]
للباب/ بالباب/ لباب/ الباب
بابه / بابها/ بابك / ...

**Reply by yes-oo** [reply]
any updates on this ? 

it's the only thing holding me back

**Reply by mzeidhassan** [reply]
@ManyTheFish just I want to explain the difference between the definite article ال which seems to be supported by Meiliserach and the following new examples submitted by @usamamhmod 

للباب in this case the root باب, a preposition ل is attached to it. 
بالباب in this case, the word الباب (with definite article), a preposition ب is attached to it. 

In the following variations, a possessive pronoun is attached to the word:

بابه meaning 'his door'
بابها meaning 'her door'
بابك meaning 'your door'

Hope this helps. 

Not sure how we can approach this issue without using a morphological analyzer, but definitely it would be an amazing feature that I don't think any other search engine has it supported yet for Arabic. Please let us know your thoughts.

**Reply by DrAliRagab** [reply]
Regarding these examples 'بابه', 'بابك', 'بابها', 'بابكم', I don't think it is required as the user will start typing with 'باب' which will show documents.
But regarding exaples 'للباب', 'بالباب', It is required to add support to this, But How??

**Reply by DrAliRagab** [reply]
Here is some prefixes with the help of ChatGPT:

## Common Arabic Prefixes

Arabic words frequently use prefixes, often combining multiple prefixes simultaneously. Below is a comprehensive list of the most common **Arabic prefixes**:

### ① Single-letter prefixes:

| Prefix | Meaning (English)        | Example (Arabic) | Example (English)  |
|--------|--------------------------|------------------|--------------------|
| **الـ** | the (definite article)   | **الكتاب**       | the book           |
| **و**   | and                      | **وكتاب**        | and a book         |
| **بـ**  | with, by, in             | **بكتاب**        | with a book        |
| **كـ**  | like, as                 | **ككتاب**        | like a book        |
| **لـ**  | for, to                  | **لكتاب**        | for a book         |
| **فـ**  | so, thus, then           | **فكتاب**        | then a book        |
| **سـ**  | will (future tense)      | **سيكتب**        | he will write      |

---

### ② Combined prefixes (two prefixes together):

| Prefix | Components | Meaning         | Arabic Example | English Example   |
|--------|------------|-----------------|----------------|-------------------|
| **بالـ** | بـ + الـ    | with/by the     | **بالقلم**     | with the pen      |
| **للـ**  | لـ + الـ    | for/to the      | **للباب**      | for the door      |
| **كالـ** | كـ + الـ    | like/as the     | **كالأسد**     | like the lion     |
| **والـ** | و + الـ     | and the         | **والقلم**     | and the pen       |
| **فالـ** | فـ + الـ    | then/so the     | **فالوقت**     | so the time       |
| **وسـ**  | و + سـ     | and will        | **وسيذهب**    | and he will go    |
| **فسـ**  | فـ + سـ     | then will       | **فسيأكل**    | then he will eat  |
| **ولـ**  | و + لـ     | and for         | **ولكتاب**    | and for a book    |
| **وبـ**  | و + بـ     | and with/by     | **وبقلم**     | and with/by a pen |
| **فلـ**  | فـ + لـ     | then for        | **فلهذا**      | then for this     |

---

### ③ Triple combinations (less frequent):

| Prefix | Components          | Meaning            | Arabic Example | English Example      |
|--------|---------------------|--------------------|----------------|----------------------|
| **وبالـ** | و + بـ + الـ         | and with/by the    | **وبالقلم**    | and with the pen     |
| **فللـ**  | فـ + لـ + الـ        | then for/to the    | **فللبيت**     | then for the house   |
| **فسـيـ** | فـ + سـ + يـ (future verb) | then he will       | **فسيذهب**    | then he will go     |
| **وللـ**  | و + لـ + الـ         | and for/to the     | **وللمدرسة**  | and for the school   |

---

### **Important Notes:**
- The prefix **"الـ"** (definite article) is the most frequent and crucial for segmentation accuracy.
- Prefixes such as **"سـ"** are primarily attached to verbs indicating future tense.
- Combination prefixes usually follow the grammatical structure: conjunction → preposition → definite article.

### **Why Prefix Handling Matters:**
- Correct prefix segmentation ensures precise indexing and retrieval in search engines.
- Improves overall accuracy and user experience for Arabic-language searches.

### **Recommendations for Implementation:**
- A comprehensive Arabic segmentation solution should handle all the above prefixes, especially common combined forms, to optimize search accuracy and relevance.

**Reply by DrAliRagab** [reply]
I think that applying all these prefixes will have a performance cost 🤔. 
Increased segmentation complexity will cause increased CPU and memory usage and slowing segmentation processing. But the main consern will be **Higher Latency in Search Results**.
Or that what I think 🤔. I am not sure.

**Reply by yes-oo** [reply]
> I think that applying all these prefixes will have a performance cost 🤔. Increased segmentation complexity will cause increased CPU and memory usage and slowing segmentation processing. But the main consern will be **Higher Latency in Search Results**. Or that what I think 🤔. I am not sure.

شكرا لكل الشرح لغير المتحدثين بالعربية 
:heart: 
and regarding to the cpu and memory usage i don't think we will get more memory usage that is noticeable, and speak about cpu yeah there will be little increase in the usage but it's worth it and it will not make a sensible difference  

again thank you for your suggestion to improve arabic in meilisearch , now i will wait for core teams to update meilisearch so i can use it in ecommerce website , finger crossed

**Reply by ManyTheFish** [reply]
hey @yes-oo @DrAliRagab!
it's an interesting subject, do we have an exhaustive list of the prefixes in Arabic somewhere?
This would be a good start to have a dictionary to better segment Arabic words. 🤔 

Thank you for your time, see you!

**Reply by DrAliRagab** [reply]
I do not think that we have such a list.

**Reply by ManyTheFish** [reply]
mmmmh 🤔 

Are there exceptions to this prefix segmentation? For instance, is there any word that starts with `الـ` but should not be segmented?

**Reply by ManyTheFish** [reply]
Hey @yes-oo and @DrAliRagab,

Do you have any use case with Meilisearch where you encounter this issue?
If so, please try adding the following setting to your index.

**PATCH**: `/indexes/:index_uid/setttings`

```json
{
"dictionary": [
  "أ",
  "ب",
  "ف",
  "ك",
  "ل",
  "و",
  "آل",
  "أب",
  "أف",
  "أك",
  "أل",
  "أو",
  "ال",
  "آلْ",
  "أَ",
  "بِ",
  "فَ",
  "فل",
  "كَ",
  "لِ",
  "لل",
  "وَ",
  "ول",
  "أبال",
  "أفال",
  "أفب",
  "أفل",
  "أكال",
  "ألل",
  "أوال",
  "أوب",
  "أول",
  "أَبِ",
  "أَفَ",
  "أَكَ",
  "أَلِ",
  "أَوَ",
  "اَلْ",
  "بال",
  "فال",
  "فَبِ",
  "فَكَ",
  "فَلِ",
  "فَلْ",
  "فلل",
  "كال",
  "لِلْ",
  "وال",
  "وَبِ",
  "وَكَ",
  "وَلِ",
  "وَلْ",
  "ولل",
  "أفبال",
  "أفلل",
  "أوبال",
  "أولل",
  "أَبِٱلْ",
  "أَفَٱلْ",
  "أَفَبِ",
  "أَفَلِ",
  "أَكَٱلْ",
  "أَلِلْ",
  "أَوَٱلْ",
  "أَوَبِ",
  "أَوَلِ",
  "بِٱلْ",
  "فَٱلْ",
  "فَبِٱلْ",
  "فَكَٱلْ",
  "فَلِلْ",
  "كَٱلْ",
  "وَٱلْ",
  "وَبِٱلْ",
  "وَكَٱلْ",
  "وَلِلْ",
  "أَفَبِٱلْ",
  "أَوَبِٱلْ",
  "أَوَلِلْ",
  "أَفَلِلْ"
]
}
```

I know it's not an optimal solution, but I'd like to see if it enhances the relevancy or if I am entirely wrong.

Thank you for your time!
